lib/LANraragi/Model/Plugins.pm

```diff
@@ -195,42 +193,43 @@ sub exec_metadata_plugin {
     my ( $plugin, $id, $oneshotarg, @args ) = @_;
     my $logger = get_logger( "Plugin System", "lanraragi" );
 
-    if ( $id eq 0 ) {
+    if ( !$id ) {
         return ( error => "Tried to call a metadata plugin without providing an id." );
     }
 
-    #If the plugin has the method "get_tags",
-    #catch all the required data and feed it to the plugin
-    if ( $plugin->can('get_tags') ) {
+    if ( !$plugin->can('get_tags') ) {
+        return ( error => "Plugin doesn't implement get_tags despite having a 'metadata' type." );
+    }
 
-        my $redis = LANraragi::Model::Config->get_redis;
-        my %hash  = $redis->hgetall($id);
+    my $redis = LANraragi::Model::Config->get_redis;
+    my %hash  = $redis->hgetall($id);
 
-        my ( $name, $title, $tags, $file, $thumbhash ) = @hash{qw(name title tags file thumbhash)};
+    my ( $name, $title, $tags, $file, $thumbhash ) = @hash{qw(name title tags file thumbhash)};
 
-        ( $_ = LANraragi::Utils::Database::redis_decode($_) ) for ( $name, $title, $tags );
+    ( $_ = LANraragi::Utils::Database::redis_decode($_) ) for ( $name, $title, $tags );
 
-        # If the thumbnail hash is empty or undefined, we'll generate it here.
-        unless ( length $thumbhash ) {
-            $logger->info("Thumbnail hash invalid, regenerating.");
-            my $thumbdir = LANraragi::Model::Config->get_thumbdir;
+    # If the thumbnail hash is empty or undefined, we'll generate it here.
+    unless ( length $thumbhash ) {
+        $logger->info("Thumbnail hash invalid, regenerating.");
+        my $thumbdir = LANraragi::Model::Config->get_thumbdir;
+        $thumbhash = "";
 
-            # Eval the thumbnail extraction, as it can error out and die
-            eval { extract_thumbnail( $thumbdir, $id, 0, 1 ) };
-            if ($@) {
-                $logger->warn("Error building thumbnail: $@");
-                $thumbhash = "";
-            } else {
-                $thumbhash = $redis->hget( $id, "thumbhash" );
-                $thumbhash = LANraragi::Utils::Database::redis_decode($thumbhash);
-            }
+        try {
+            extract_thumbnail( $thumbdir, $id, 0, 1 );
+            $thumbhash = $redis->hget( $id, "thumbhash" );
+            $thumbhash = LANraragi::Utils::Database::redis_decode($thumbhash);
+        } catch ($e) {
+            $logger->warn("Error building thumbnail: $e");
         }
-        $redis->quit();
+    }
+    $redis->quit();
 
+    my %returnhash;
+    try {
         # Hand it off to the plugin here.
         # If the plugin requires a login, execute that first to get a UserAgent
         my %pluginfo = $plugin->plugin_info();
-        my $ua       = exec_login_plugin( $pluginfo{login_from} );
+        my $ua       = exec_login_plugin( $pluginfo{login_from} ) if ( exists $pluginfo{login_from} );
```

Comment:
I believe this would make `$ua` undef if there's no login plugin set.  
This is technically different from the old behavior where you would always have a Mojo::UserAgent object available in `%infohash`, even if you don't depend on a login plugin.  

I don't know if external plugins came to depend on this, but it might be safer to keep the old behavior here.

Reply:
I don't know why at that time I was really sure that it was created here only for plugins that required to login. Shame on me!